File: YukariLauncher/src/main/java/net/kdt/pojavlaunch/customcontrols/gamepad/GamepadEmulatedButton.java
package net.kdt.pojavlaunch.customcontrols.gamepad;

import android.view.KeyEvent;

/**
 * このクラスは、ゲームパッドに物理的には存在しないが、
 * emulated from other inputs on it (like WASD directional keys)
 */
public class GamepadEmulatedButton {
    public short[] keycodes;
    protected boolean mIsDown = false;
/**
 * 「update」メソッド。
 * このクラスに定義された機能メソッドです。
 */
    public void update(KeyEvent event) {
        boolean isKeyDown = (event.getAction() == KeyEvent.ACTION_DOWN);
        update(isKeyDown);
    }
/**
 * 「update」メソッド。
 * このクラスに定義された機能メソッドです。
 */
    public void update(boolean isKeyDown){
        if(isKeyDown != mIsDown){
            mIsDown = isKeyDown;
            onDownStateChanged(mIsDown);
        }
    }
/**
 * 「reset Button State」処理を実行します。
 * このメソッドは特定の機能を提供するために実装されています。
 */
    public void resetButtonState() {
        if(mIsDown) Gamepad.sendInput(keycodes, false);
        mIsDown = false;
    }
/**
 * 「on Down State Changed」処理を実行します。
 * このメソッドは特定の機能を提供するために実装されています。
 */
    protected void onDownStateChanged(boolean isDown) {
        Gamepad.sendInput(keycodes, mIsDown);
    }
}

File: YukariLauncher/src/main/java/com/arata/yukarilauncher/ui/subassembly/customcontrols/gamepad/GamepadEmulatedButton.kt
package com.arata.yukarilauncher.ui.subassembly.customcontrols.gamepad

import android.view.KeyEvent

/**
 * ゲームパッドのエミュレートされたボタンを表すクラス。
 * 物理ボタン入力をMinecraftのキーコードに変換して送信します。
 */
open class GamepadEmulatedButton {
    /** このボタンに割り当てられたキーコード配列。 */
    var keycodes: ShortArray = shortArrayOf()
    protected var mIsDown = false

    /**
     * KeyEventからボタン状態を更新します。
     * @param event 入力イベント
     */
    fun update(event: KeyEvent) {
        val isKeyDown = event.action == KeyEvent.ACTION_DOWN
        update(isKeyDown)
    }

    /**
     * 押下状態を更新します。
     * @param isKeyDown 押下されているかどうか
     */
    fun update(isKeyDown: Boolean) {
        if (isKeyDown != mIsDown) {
            mIsDown = isKeyDown
            onDownStateChanged(mIsDown)
        }
    }

    /** ボタン状態をリセットします。 */
    open fun resetButtonState() {
        if (mIsDown) Gamepad.sendInput(keycodes, false)
        mIsDown = false
    }

    /**
     * 押下状態が変更されたときに呼ばれます。
     * @param isDown 現在の押下状態
     */
    protected open fun onDownStateChanged(isDown: Boolean) {
        Gamepad.sendInput(keycodes, mIsDown)
    }
}